src/Microsoft.ML.Dnn/ImageClassificationTransform.cs

```diff
@@ -981,6 +998,108 @@ public sealed class BottleneckMetrics
+            /// Current number of epochs where there has been no improvement.
+            /// Stop training when wait >=patience.
+            /// </summary>
+            private int _wait;
+
+            /// <summary>
+            /// The metric to be monitored (eg Accuracy, Loss).
+            /// </summary>
+            private EarlyStoppingMetric _metric;
+
+            /// <summary>
+            /// Minimum change in the monitored quantity to be considered as an improvement.
+            /// </summary>
+            public float MinDelta { get; set; }
+
+            /// <summary>
+            /// Number of epochs to wait after no improvement is seen consecutively
+            /// before stopping the training.
+            /// </summary>
+            public int Patience { get; set; }
+
+            /// <summary>
+            /// Whether the monitored quantity is to be increasing (eg. Accuracy, CheckIncreasing = true)
+            /// or decreasing (eg. Loss, CheckIncreasing = false).
+            /// </summary>
+            public bool CheckIncreasing { get; set; }
+
+            public EarlyStopping(float minDelta = 0.01f, int patience = 20, EarlyStoppingMetric metric = EarlyStoppingMetric.Accuracy, bool checkIncreasing = true)
+            {
+                _bestMetricValue = 0.0f;
+                _wait = 0;
+                _metric = metric;
+                MinDelta = Math.Abs(minDelta);
+                Patience = patience;
+                CheckIncreasing = checkIncreasing;
+
+                //Set the CheckIncreasing according to the metric being monitored
+                if (metric == EarlyStoppingMetric.Accuracy)
+                    CheckIncreasing = true;
+                else if (metric == EarlyStoppingMetric.Loss)
+                    CheckIncreasing = false;
+            }
+
+            /// <summary>
+            /// To be called at the end of every epoch to check if training should stop.
+            /// For increasing metric(eg.: Accuracy), if metric stops increasing, stop training if
+            /// value of metric doesn't increase within 'patience' number of epochs.
+            /// For decreasing metric(eg.: Loss), stop training if value of metric doesn't decrease
+            /// within 'patience' number of epochs.
+            /// Any change  in the value of metric of less than 'minDelta' is not considered a change.
+            /// </summary>
+            public bool ShouldStop(TrainMetrics currentMetrics)
+            {
+                float currentMetricValue;
+                if (_metric == EarlyStoppingMetric.Accuracy)
+                    currentMetricValue = currentMetrics.Accuracy;
+                else
+                    currentMetricValue = currentMetrics.CrossEntropy;
+                if(CheckIncreasing)
+                {
+                    if((currentMetricValue- _bestMetricValue) < MinDelta)
+                    {
+                        _wait += 1;
+                        if(_wait >= Patience)
```

Comment:
>if(_wait >= Patience) [](start = 24, length = 21)

Can _wait ever be greater than Patience?

Reply:
since patience is an int, user might supply a negative value. in that case, it is better to check >= instead of ==.

---
In reply to: [330317534](https://github.com/dotnet/machinelearning/pull/4237#discussion_r330317534) [](ancestors = 330317534)